src/lib/netlist/core/core_device.h

```diff
@@ -33,7 +33,7 @@ namespace netlist
 		{
 			gsl_Expects(m_active_outputs >= 0);
 
-			if (m_activate && m_hint_deactivate)
+			if (!m_activate.isnull() && m_hint_deactivate)
```

Comment:
Is there a good reason for avoiding the use of an `explicit operator bool`?  It would be the idiomatic C++ approach for testing if the object is dereferenceable, as used with pointers, various smart pointers (`std::unique_ptr`, `std::shared_ptr`, etc.), `std::optional`, `std::function`, and so on.

Reply:
I have removed the bool opeator to align the design with `delegate` from `delegate.h`. `delegate`does not define an `explicit operator bool`.

Reply:
Yeah, but that isn’t necessarily a good thing.  MAME’s `delegate` lacks an `explicit operator bool` for two reasons:
* Before C++11, if you had `operator bool` you could end up accidentally casting to integer types when you didn’t want to (leading to developer frustration), and if you had `explicit operator bool` you needed to explicitly cast to `bool` in places like control statement conditions.  It was often considered more trouble than it was worth.  C++11 changed the rules so that in certain contexts, the `explicit operator bool` will be used automatically, making it easier to work with.  The `delegate` predates C++11 and hasn’t had an `explicit operator bool` added since then.
* I haven’t just added `explicit operator bool` as a synonym for `isnull` because I haven’t checked the exact behaviour of `isnull()` when late binding is involved.  I wouldn’t want to add an `explicit operator bool` with confusing semantics.
* Some people seem to have an aversion to every using `operator bool` at all, and won’t use it even when it’s the idiomatic way to test something.

If you can implement `explicit operator bool` for PPMF in an intuitive way, I wouldn’t consider `delegate` lacking an equivalent to be a good reason not to.

Reply:
I removed it. No problem adding it again. Would be great if I can do this in a separate PR since I have other stuff I'd like to commit and waiting for this PR to be applied.

Reply:
> Would be great if I can do this in a separate PR since I have other stuff I'd like to commit and waiting for this PR to be applied.

I’d still like to see this actually tested on Itanium ARM variant, since that’s a key part of #8590.  Any of the following will do:
* An M1 Mac
* Emscripten build environment on Linux or Windows (can netlist library be built standalone with emmake wrapper?  e.g. see [how MAME is built for WebAssembly](https://docs.mamedev.org/initialsetup/compilingmame.html#emscripten-javascript-and-html))
* If the PPMF tests can be isolated so the rest of the netlist stuff doesn’t need to be built, a Raspberry Pi running Linux, or even one of those Chinese MIPS/Linux hand-helds sold as emulation devices

Reply:
Here we go. I'll commit the changes to support emscripten with `make emsdk` in a second:

```
emrun  --browser chrome nltool.html -c tests
Starting browser: /usr/bin/google-chrome-stable --enable-nacl --enable-pnacl --disable-restore-session-state --enable-webgl --no-default-browser-check --no-first-run --allow-file-access-from-files http://localhost:6931/nltool.html?-c&tests
======================================
Running 9 tests
======================================
pstring::conversion:
	OK: putf8string("Общая ком") == putf8string(putf16string(putf8string("Общая ком")))
	OK: putf8string("Общая ком") == putf8string(putf16string("Общая ком"))
pfunction::operators:
	OK: 1.0 == plib::pfunction<double>("1==1")()
	OK: 1.0 == plib::pfunction<double>("1 *0 == 2-1-1")()
	OK: 1.0 == plib::pfunction<double>("0!=1")()
	OK: 1.0 == plib::pfunction<double>("0<1")()
	OK: 1.0 == plib::pfunction<double>("1>0")()
	OK: 1.0 == plib::pfunction<double>("0<=1")()
	OK: 1.0 == plib::pfunction<double>("1>=0")()
	OK: 1.0 == plib::pfunction<double>("1<=1")()
	OK: 1.0 == plib::pfunction<double>("1>=1")()
	OK: 1.0 == plib::pfunction<double>("0!=a", {"a"})({1.0})
pfunction::func_if:
	OK: 2.0 == plib::pfunction<double>("if(1>0, 2, 0)")()
	OK: 3.0 == plib::pfunction<double>("if(0>1, 2, 3)")()
	OK: 2.0 == plib::pfunction<double>("if(sin(1)>0, 2, 3)")()
	OK: 1.0 == plib::pfunction<double>("if(A2>2.5, 0-A1, (0.07-(0.005*A1))*if(A0>2.5,1,0-1))", {"A0","A1","A2"})({1.0,-1.0,3.0})
	OK: -0.065 == plib::pfunction<double>("if(A2>2.5, 0-A1, (0.07-(0.005*A1))*if(A0>2.5,1,0-1))", {"A0","A1","A2"})({1.0,1.0,1.0})
	OK: 0.065 == plib::pfunction<double>("if(A2>2.5, 0-A1, (0.07-(0.005*A1))*if(A0>2.5,1,0-1))", {"A0","A1","A2"})({3.0,1.0,1.0})
	OK: 1.0 == plib::pfunction<double>("0!=a", {"a"})({1.0}) == true
pfunction::unary_minus:
	OK: 1.0 == plib::pfunction<double>("-1>-2")()
	OK: 12.0 == plib::pfunction<double>("(-3)*(-4)")()
	OK: 12.0 == plib::pfunction<double>("(-3)*-4")()
	OK: 12.0 == plib::pfunction<double>("-3*-4")()
	OK: -3.0 == plib::pfunction<double>("-A0", {"A0"})({3.0})
	OK: -9.0 == plib::pfunction<double>("3*-trunc(3.2)")()
	OK: -18.0 == plib::pfunction<double>("3*-(3*2)")()
	OK: -12.0 == plib::pfunction<double>("3*-(2*1)^2")()
	OK: got no exception for plib::pfunction<double>("(-3)")()
pfunction::expect_throw:
	OK: got plib::pexception for plib::pfunction<double>("(3, 4)")()
	OK: got plib::pexception for plib::pfunction<double>("((3)")()
	OK: got plib::pexception for plib::pfunction<double>("(3))")()
pmfp_test_complex_return::multibase_test:
	OK: f() == ret_t(static_cast<void *>(static_cast<a *>(&obj)), 1)
	OK: g_ret == ret_t(static_cast<void *>(static_cast<b *>(&obj)), 2)
	OK: g_ret == h()
	OK: g_ret == f()
	OK: f1_ret.first == f2_ret.first
	OK: f1_ret.second == 3
	OK: f2_ret.second == 4
	OK: f1_ret.first == f2_ret.first
	OK: f1_ret.second == 13
	OK: f2_ret.second == 14
	OK: static_cast<void *>(static_cast<vb *>(&obj2)) != static_cast<void *>(static_cast<vd1 *>(&obj2))
	OK: static_cast<void *>(static_cast<vb *>(&obj2)) != static_cast<void *>(static_cast<vd2 *>(&obj2))
	OK: f().second == 8
	OK: f().second == 9
pmfp_test_simple_return::multibase_test:
	OK: fr == ret_t(static_cast<void *>(static_cast<a *>(&obj)), 1)
	OK: gr == ret_t(static_cast<void *>(static_cast<b *>(&obj)), 2)
	OK: gr == hr
	OK: gr == fr
	OK: f1_ret.first == f2_ret.first
	OK: f1_ret.second == 3
	OK: f2_ret.second == 4
	OK: f1_ret.first == f2_ret.first
	OK: f1_ret.second == 13
	OK: f2_ret.second == 14
	OK: static_cast<void *>(static_cast<vb *>(&obj2)) != static_cast<void *>(static_cast<vd1 *>(&obj2))
	OK: static_cast<void *>(static_cast<vb *>(&obj2)) != static_cast<void *>(static_cast<vd2 *>(&obj2))
	OK: fr.second == 8
	OK: fr.second == 9
test_late_pmfp::late_pmfp:
	OK: r == 1 == true
test_compile::compile:
	OK: mfp_PPMF_TYPE_PMF.isnull() == true
	OK: mfp_PPMF_TYPE_INTERNAL_ITANIUM.isnull() == true
	OK: mfp_PPMF_TYPE_INTERNAL_ARM.isnull() == true
	OK: mfp_PPMF_TYPE_INTERNAL_MSC.isnull() == true
	OK: "PPMF_TYPE_GNUC_PMF_CONV not supported on this build" != ""
	OK: plib::ppmf_internal::value == 3
```
All tests are successful. The last line shows that PPMF_TYPE_INTERNAL_ARM is used by default.